### PR TITLE
Inject ILogger to RequestLocalization Middleware

### PR DESCRIPTION
**Pull request by Kahbazi**
https://github.com/aspnet/AspNetCore/issues/6532

### CODE REVIEW

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs [hunk 1]
```diff
@@ -26,22 +26,24 @@ public class RequestLocalizationMiddleware
...
 
         /// <summary>
         /// Creates a new <see cref="RequestLocalizationMiddleware"/>.
         /// </summary>
         /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
         /// <param name="options">The <see cref="RequestLocalizationOptions"/> representing the options for the
+        /// <param name="loggerFactory">The <see cref="ILoggerFactory"/> used for logging.</param>
         /// <see cref="RequestLocalizationMiddleware"/>.</param>
-        public RequestLocalizationMiddleware(RequestDelegate next, IOptions<RequestLocalizationOptions> options)
+        public RequestLocalizationMiddleware(RequestDelegate next, IOptions<RequestLocalizationOptions> options, ILoggerFactory loggerFactory)
```
Comment (gfoidl) [inline src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: Why did you take the change in https://github.com/aspnet/AspNetCore/pull/10946/commits/16739e0cfdbff8c4b50ea3398da4b7d9f585c561?

So to not inject the `ILogger<RequestLocalizationMiddleware>`.

Comment (Kahbazi) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: I saw that other middlewares inject `ILoggerFactory` so I wanted to keep the same pattern

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs [hunk 2]
```diff
@@ -26,22 +26,24 @@ public class RequestLocalizationMiddleware
...
         {
             if (options == null)
             {
                 throw new ArgumentNullException(nameof(options));
             }
 
             _next = next ?? throw new ArgumentNullException(nameof(next));
+            _logger = loggerFactory?.CreateLogger<RequestLocalizationMiddleware>() ?? throw new ArgumentNullException(nameof(next));
```
Comment (Tratcher) [inline src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]:
```suggestion
            _logger = loggerFactory?.CreateLogger<RequestLocalizationMiddleware>() ?? throw new ArgumentNullException(nameof(loggerFactory));
```

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs (hunk 1 above)
Comment (Tratcher) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: The breaking change guidelines are to add a new constructor and obsolete the old one. For weird DI reasons the new constructor should be placed above the old one.

Comment (Kahbazi) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: So how should we create logger object for the old constructor. Maybe I should put back the EnsureLogger!!?

Comment (Kahbazi) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: So how should we create logger object for the old constructor. Maybe I should put back the EnsureLogger!!?

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs (hunk 2 above)
Comment (Kahbazi) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: Oops

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs (hunk 1 above)
Comment (rynowak) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: > For weird DI reasons the new constructor should be placed above the old one.

Does this actually matter to DI? the order used by reflection is not a stable guarantee 😢

Comment (Tratcher) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: It does for the ActivatorUtilities library we use to construct middleware. Middleware aren't in the DI container so we build them manually and pull their dependencies from DI.

Comment (pranavkm) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: There's an attribute that you can use to specify the constructor to be used - `ActivatorUtilitiesConstructorAttribute`

Comment (pranavkm) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: @Kahbazi here's what I'd do:

1) Add a new ctor which accepts a `ILoggerFactory` \ `ILogger<T>` instance. I personally prefer the latter, but there isn't a clear guidance for one over the other.
2) Add `[Obsolete]` attribute to the current attribute. Add `ActivatorUtilitiesConstructorAttribute` to the new ctor
3) Change the current constructor to call the other ctor. Have it pass a `NullLogger` instance. This should avoid the need to have `EnsureLogger`

Comment (Tratcher) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: @Kahbazi here's an example, we use a NullLogger for the other constructor:
https://github.com/aspnet/AspNetCore/blob/62351067ff4c1401556725b401478e648b66acdc/src/Security/CookiePolicy/src/CookiePolicyMiddleware.cs#L20-L32

Comment (Kahbazi) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: So the log will be disabled if someone directly uses the old constructor, is that ok?

Comment (Tratcher) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: Yes, it's quite unlikely anyone will use the old one, but we leave it for binary compat.

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs [hunk 3]
```diff
@@ -26,14 +27,35 @@ public class RequestLocalizationMiddleware
...
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="RequestLocalizationMiddleware"/>.
+        /// </summary>
+        /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
+        /// <param name="options">The <see cref="RequestLocalizationOptions"/> representing the options for the
+        /// <see cref="RequestLocalizationMiddleware"/>.</param>
+        [Obsolete("This is obsolete and will be removed in a future version. Use RequestLocalizationMiddleware(RequestDelegate next, IOptions<RequestLocalizationOptions> options, ILoggerFactory loggerFactory) instead")]
         public RequestLocalizationMiddleware(RequestDelegate next, IOptions<RequestLocalizationOptions> options)
```
Comment (pranavkm) [inline src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: Could you update this ctor to call the other one:

```C#
public RequestLocalizationMiddleware(RequestDelegate next, IOptions<RequestLocalizationOptions> options)
   : this(next, options, NullLoggerFactory.Instance)
```

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs [hunk 4]
```diff
@@ -26,14 +27,35 @@ public class RequestLocalizationMiddleware
...
+            _options = options.Value;
+        }
+
+        /// <summary>
+        /// Creates a new <see cref="RequestLocalizationMiddleware"/>.
+        /// </summary>
+        /// <param name="next">The <see cref="RequestDelegate"/> representing the next middleware in the pipeline.</param>
+        /// <param name="options">The <see cref="RequestLocalizationOptions"/> representing the options for the
+        /// <see cref="RequestLocalizationMiddleware"/>.</param>
+        [Obsolete("This is obsolete and will be removed in a future version. Use RequestLocalizationMiddleware(RequestDelegate next, IOptions<RequestLocalizationOptions> options, ILoggerFactory loggerFactory) instead")]
```
Comment (pranavkm) [inline src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: `[Obsolete("This constructor is obsolete and will be removed in a future version. Use RequestLocalizationMiddleware(RequestDelegate next, IOptions<RequestLocalizationOptions> options, ILoggerFactory loggerFactory) instead")]`

File: src/Middleware/Localization/src/RequestLocalizationMiddleware.cs (hunk 3 above)
Comment (Kahbazi) [reply src/Middleware/Localization/src/RequestLocalizationMiddleware.cs]: Done

**Comment by Tratcher** [comment]
So far so good. @pranavkm can you track merging this?

**Comment by pranavkm** [comment]
@Kahbazi 👍 

```
src\Middleware\Localization\ref\Microsoft.AspNetCore.Localization.netcoreapp3.0.cs(,): error : Generated code is not up to date in src/Middleware/Localization/ref/Microsoft.AspNetCore.Localization.netcoreapp3.0.cs. You might need to regenerate the reference assemblies or project list (see docs/ReferenceAssemblies.md and docs/ReferenceResolution.md)
PowerShell exited with code '1'.
```

**Comment by Kahbazi** [comment]
@pranavkm Done!

**Comment by pranavkm** [comment]
Thanks for the PR!

**Pull request merged**